runtime/src/tiered_storage/file.rs

```diff
@@ -0,0 +1,73 @@
+use std::{
+    fs::{File, OpenOptions},
+    io::{Read, Seek, SeekFrom, Write},
+    mem,
+    path::Path,
+};
+
+#[derive(Debug)]
+pub struct TieredStorageFile {
+    pub file: File,
+}
+
+impl TieredStorageFile {
+    /// Creates a tiered-storage file.
+    /// If the create flag is false, it will open an existing file
+    /// in read-only mode.
+    pub fn new<P: AsRef<Path>>(file_path: P, create: bool) -> Self {
+        let file = OpenOptions::new()
+            .read(true)
+            .write(create)
+            .create(create)
+            .open(file_path.as_ref())
+            .map_err(|e| {
+                panic!(
+                    "Unable to {} data file {} in current dir({:?}): {:?}",
+                    if create { "create" } else { "open" },
+                    file_path.as_ref().display(),
+                    std::env::current_dir(),
+                    e
+                );
+            })
+            .unwrap();
+        Self { file }
+    }
+
+    pub fn write_type<T>(&self, value: &T) -> Result<usize, std::io::Error> {
+        unsafe {
+            let ptr =
+                std::slice::from_raw_parts((value as *const T) as *const u8, mem::size_of::<T>());
+            (&self.file).write_all(ptr)?;
+        }
+        Ok(std::mem::size_of::<T>())
+    }
+
+    pub fn read_type<T>(&self, value: &mut T) -> Result<(), std::io::Error> {
+        unsafe {
+            let ptr =
+                std::slice::from_raw_parts_mut((value as *mut T) as *mut u8, mem::size_of::<T>());
+            (&self.file).read_exact(ptr)?;
+        }
+        Ok(())
+    }
+
+    pub fn seek(&self, offset: u64) -> Result<u64, std::io::Error> {
+        (&self.file).seek(SeekFrom::Start(offset))
```

Comment:
Was poking around, and do not understand this actually, but it corresponds to point 2 from your other comment: https://github.com/solana-labs/solana/pull/31161/files#r1167050134

Having `(&self.file).XXX` instead of `self.file.XXX`, let's it call mutable functions of `File`. Removing the parenthesis and `&` will make this fail with compile error:

```rust
error[E0596]: cannot borrow `self.file` as mutable, as it is behind a `&` reference
  --> runtime/src/tiered_storage/file.rs:40:9
   |
40 |         self.file.write_all(ptr)?;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^ `self` is a `&` reference, so the data it refers to cannot be borrowed as mutable
   |
help: consider changing this to be a mutable reference
   |
36 |     pub fn write_type<T>(&mut self, value: &T) -> Result<usize, std::io::Error> {
   |                          ~~~~~~~~~
```

Reply:
Oh wow, that's very interesting....the implementation of `Write` is **also** done for `&File` in the standard library:
```rust
#[stable(feature = "rust1", since = "1.0.0")]
impl Write for &File {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        self.inner.write(buf)
    }

    fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
        self.inner.write_vectored(bufs)
    }

    #[inline]
    fn is_write_vectored(&self) -> bool {
        self.inner.is_write_vectored()
    }

    fn flush(&mut self) -> io::Result<()> {
        self.inner.flush()
    }
}
```

So `(&self.file)` is giving us a `&mut &File`, which implements these traits.

Reply:
@apfitzge: Thanks for forwarding the detailed answer while I was previously OOO.

> I guess then should we go the current route, or would it make more sense to do &mut self for the write/modifying functions?

We should go the current route.  Basically, you can think of this as something like an instance of MmapMut in AppendVec.  If we do the &mut self route, then append_accounts or other functions also need to take `&mut self`, which will all-the-way propagate the change to AccountStorageEntry and many other accounts_db code.